File: src/main/java/com/mpe85/grampa/input/impl/CharSequenceLineCounter.java
package com.mpe85.grampa.input.impl;

import java.util.NavigableMap;
import java.util.Optional;
import java.util.TreeMap;

import com.google.common.base.Preconditions;
import com.mpe85.grampa.input.InputPosition;
import com.mpe85.grampa.input.LineCounter;

/**
 * A {@link LineCounter} implementation for {@link CharSequence}s.
 * 
 * @author mpe85
 *
 */
public class CharSequenceLineCounter implements LineCounter {
	
	public CharSequenceLineCounter(final CharSequence input) {
		Preconditions.checkNotNull(input, "An 'input' must not be null.");
		length = input.length();
		initLines(input);
	}
	
	private void initLines(final CharSequence input) {
		for (int currentIdx = 0, lineStartIdx = 0, lineNumber = 0; currentIdx < length; currentIdx++) {
			if (input.charAt(currentIdx) == LF) {
				lines.put(lineStartIdx, ++lineNumber);
				lineStartIdx = currentIdx + 1;
			}
			else if (currentIdx == length - 1) {
				lines.put(lineStartIdx, ++lineNumber);
			}
		}
	}
	
	@Override
	public int getLineCount() {
		return lines.size();
	}
	
	@Override
	public InputPosition getPosition(final int index) {
		Preconditions.checkElementIndex(index, length, "An 'index' must not be out of range.");
		return Optional.of(index)
				.map(lines::floorEntry)
				.map(e -> new InputPosition(e.getValue(), index - e.getKey() + 1))
				.get();
	}
	
	
	private static final char LF = '\n';
	
	private final NavigableMap<Integer, Integer> lines = new TreeMap<>();
	private final int length;
	
}

File: src/main/kotlin/com/mpe85/grampa/input/impl/CharSequenceLineCounter.kt
package com.mpe85.grampa.input.impl

import com.mpe85.grampa.input.InputPosition
import com.mpe85.grampa.input.LineCounter
import java.util.*
import kotlin.streams.asSequence

/**
 * A [LineCounter] implementation for [CharSequence]s.
 *
 * @author mpe85
 * @property[input] The input in which the lines are counted
 */
class CharSequenceLineCounter(private val input: CharSequence) : LineCounter {

    override val lineCount get() = lines.size
    private val lines = getLines(input)


    private fun getLines(input: CharSequence): NavigableMap<Int, Int> {
        val map = TreeMap<Int, Int>()
        var lineStartIdx = 0
        var lineNumber = 0
        input.chars().asSequence().forEachIndexed { index, ch ->
            if (ch == LF) {
                map[lineStartIdx] = ++lineNumber
                lineStartIdx = index + 1
            } else if (index == input.length - 1) {
                map[lineStartIdx] = ++lineNumber
            }
        }
        return map
    }

    override fun getPosition(index: Int) = lines.floorEntry(checkBounds(index)).let {
        InputPosition(it.value, index - it.key + 1)
    }

    private fun checkBounds(index: Int) = index.also { input[it] }

    companion object {
        private const val LF: Int = '\n'.toInt()
    }

}
